source/extensions/filters/http/oauth/oauth_client.cc

```diff
@@ -0,0 +1,148 @@
+
+void OAuth2ClientImpl::asyncGetAccessToken(const std::string& auth_code,
+                                           const std::string& client_id, const std::string& secret,
+                                           const std::string& cb_url) {
+  Http::RequestMessagePtr request = createPostRequest();
+  request->headers().setPath(AuthTokenEndpoint);
+  const std::string body =
+      fmt::format(GetAccessTokenBodyFormatString, auth_code, client_id, secret, cb_url);
+  request->body() = std::make_unique<Buffer::OwnedImpl>(body);
+
+  ENVOY_LOG(debug, "Dispatching OAuth request for access token.");
+  dispatchRequest(std::move(request));
+  state_ = OAuthState::PendingAccessToken;
+}
+
+void OAuth2ClientImpl::asyncGetIdentity(const std::string& access_token) {
+  Http::RequestMessagePtr request = createAuthGetRequest(access_token);
+  request->headers().setPath(AuthMeApiEndpoint);
+
+  ENVOY_LOG(debug, "Dispatching OAuth request for user identity.");
+  dispatchRequest(std::move(request));
+  state_ = OAuthState::PendingIdentity;
+}
+
+void OAuth2ClientImpl::dispatchRequest(Http::RequestMessagePtr&& msg) {
+  in_flight_request_ = cm_.httpAsyncClientForCluster(cluster_name_)
+                           .send(std::move(msg), *this,
+                                 Http::AsyncClient::RequestOptions().setTimeout(timeout_duration_));
+}
+
+// successful request calls will end up here
+void OAuth2ClientImpl::onSuccess(const Http::AsyncClient::Request&,
+                                 Http::ResponseMessagePtr&& message) {
+  in_flight_request_ = nullptr;
+
+  /**
+   * Due to the asynchonous nature of this client, it's important to immediately update the state
+   * that we are not waiting on anything, otherwise a painful debugging session may ensue.
+   */
+  OAuthState prior_state = state_;
+  state_ = OAuthState::Idle;
+
+  // Check that the auth cluster returned a happy response.
+  const auto response_code = message->headers().Status()->value().getStringView();
+  if (response_code != "200") {
+    ENVOY_LOG(debug, "Oauth response code: {}", response_code);
+    parent_->sendUnauthorizedResponse();
+    return;
+  }
+
+  const std::string response_body = message->bodyAsString();
+
+  Json::ObjectSharedPtr json_object;
+  try {
+    json_object = Json::Factory::loadFromString(response_body);
+  } catch (const Json::Exception& e) {
+    ENVOY_LOG(debug, "Error parsing response body, received JSON Exception: {}", e.what());
+    ENVOY_LOG(debug, "Response body: {}", response_body);
+    parent_->sendUnauthorizedResponse();
+    return;
+  }
+  if (json_object == nullptr) {
+    ENVOY_LOG(debug, "No json body was loaded.");
+    parent_->sendUnauthorizedResponse();
+    return;
+  }
+  // the current state of the client dictates how we interpret the response
+  if (prior_state == OAuthState::PendingAccessToken) {
+    if (!(json_object->hasObject(kAccessToken) && json_object->hasObject(kExpiresIn))) {
```

Comment:
de morgan this? thats the expression used by the log line, seems like it would be a bit more readable